### PR TITLE
Classroom reservations

### PR DESCRIPTION
**Pull request by AngleQian**

### CODE REVIEW

File: oclubs-tables.sql [hunk 1]
```diff
@@ -61,13 +64,15 @@ CREATE TABLE activity (
 	act_location varchar(255) NOT NULL,
 	act_cas int NOT NULL, # CAS hours
 	act_post int NOT NULL, # Foreign key to text.text_id
-	act_selections varchar(255) NOT NULL # stores object in JSON
+	act_selections varchar(255), NOT NULL # stores object in JSON
```
Comment (zhuyifei1999) [inline oclubs-tables.sql]: Syntax error, comma after NOT NULL

File: oclubs-tables.sql [hunk 2]
```diff
@@ -157,3 +162,37 @@ CREATE TABLE preferences (
...
+	room_studentsToUseLunch boolean NOT NULL, # true = available to students, false = not available to students
+	room_studentsToUseAfternoon boolean NOT NULL, # true = available to students, false = not available to students
+	room_building tinyint, # Enum Building
+	room_desc varchar(255) # optional descriptors (eg ASB only)
+);
+
+CREATE INDEX room_id ON classroom (room_id);
+
+CREATE TABLE reservation (
+	res_id int unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT;
```
Comment (zhuyifei1999) [inline oclubs-tables.sql]: Syntax error. Comma not semicolon

File: oclubs-tables.sql [hunk 3]
```diff
@@ -157,3 +162,37 @@ CREATE TABLE preferences (
...
+CREATE TABLE classroom (
+	room_id int unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT,
+	room_number varchar(16) NOT NULL,
+	room_studentsToUseLunch boolean NOT NULL, # true = available to students, false = not available to students
+	room_studentsToUseAfternoon boolean NOT NULL, # true = available to students, false = not available to students
+	room_building tinyint, # Enum Building
+	room_desc varchar(255) # optional descriptors (eg ASB only)
+);
+
+CREATE INDEX room_id ON classroom (room_id);
```
Comment (zhuyifei1999) [inline oclubs-tables.sql]: You don't need an explicit index on primary key.

File: oclubs-tables.sql [hunk 4]
```diff
@@ -157,3 +162,37 @@ CREATE TABLE preferences (
...
+	res_owner int, # Foreign key to user.user_id
+	res_classroom int NOT NULL, # Foreign key to classroom.room_id
+	res_SBNeeded boolean NOT NULL, # true = need smartboard, false = no need smartboard
+	res_SBAppDesc varchar(512),
+	res_instructors_approval boolean DEFAULT false,
+	res_directors_approval boolean DEFAULT false,
+	res_SBApp_status tinyint DEFAULT 0 # based on SBAppStatus enum
+);
+
+CREATE INDEX res_id ON classroom (res_id);
```
Comment (zhuyifei1999) [inline oclubs-tables.sql]: Ditto. Also the table name is incorrect

File: oclubs-tables.sql [hunk 5]
```diff
@@ -157,3 +162,37 @@ CREATE TABLE preferences (
 );
 
 CREATE INDEX pref_user ON preferences (pref_user);
+
+CREATE TABLE classroom (
+	room_id int unsigned NOT NULL PRIMARY KEY AUTO_INCREMENT,
+	room_number varchar(16) NOT NULL,
+	room_studentsToUseLunch boolean NOT NULL, # true = available to students, false = not available to students
+	room_studentsToUseAfternoon boolean NOT NULL, # true = available to students, false = not available to students
+	room_building tinyint, # Enum Building
```
Comment (zhuyifei1999) [inline oclubs-tables.sql]: NOT NULL

File: oclubs-tables.sql [hunk 6]
```diff
@@ -157,3 +162,37 @@ CREATE TABLE preferences (
...
+	res_timeslot tinyint NOT NULL, # based on ActivityTime enum
+	res_status tinyint NOT NULL, # 1 = club not yet paired with activity, # 2 = club paired with activity,  # 3 = non-club
+	res_activity_name varchar(255),
+	res_reserver_name varchar(255),
+	res_reserver_club int, # Foreign key to club.club_id
+	res_owner int, # Foreign key to user.user_id
+	res_classroom int NOT NULL, # Foreign key to classroom.room_id
+	res_SBNeeded boolean NOT NULL, # true = need smartboard, false = no need smartboard
+	res_SBAppDesc varchar(512),
+	res_instructors_approval boolean DEFAULT false,
```
Comment (zhuyifei1999) [inline oclubs-tables.sql]: NOT NULL, and also for the two columns below.

File: oclubs/access/database.py [hunk 7]
```diff
@@ -453,3 +453,28 @@ def delete_rows(table, conds):
         raise NoRow
 
     return rows
+
+
+def delete_rows_multiple(table, conds):
+    """
+    Forms correct syntax for DELETE, supports JOIN
+    TODO: Merge functionality with delete_rows
```
Comment (zhuyifei1999) [inline oclubs/access/database.py]: Yes please do this.

File: oclubs/blueprints/actblueprint.py [hunk 8]
```diff
@@ -11,21 +11,23 @@
...
+from oclubs.objs import User, Activity, Upload, FormattedText, Reservation
 from oclubs.exceptions import UploadNotSupported, NoRow
+from oclubs.forms.reservation_forms import PairReservation
 
 actblueprint = Blueprint('actblueprint', __name__)
 
 
-@actblueprint.route('/all/<clubfilter:club_filter>/', defaults={'page': 1})
-@actblueprint.route('/all/<clubfilter:club_filter>/<int:page>')
+@actblueprint.route('viewlist/<clubfilter:club_filter>/', defaults={'page': 1})
```
Comment (zhuyifei1999) [inline oclubs/blueprints/actblueprint.py]: Prefix viewlist with /

File: oclubs/blueprints/actblueprint.py [hunk 9]
```diff
@@ -302,8 +316,64 @@ def changeactinfo_submit(activity):
...
+
+    form = PairReservation()
+
+    if reservations:
+        choices = []
+        for reservation in reservations:
+            if reservation.activity is None:
+                choices.append((
+                    reservation.callsign,
+                    str(reservation.date) + ' ' +
```
Comment (zhuyifei1999) [inline oclubs/blueprints/actblueprint.py]: Use ''.format() or '' % (...)

Comment (zhuyifei1999) [reply oclubs/blueprints/actblueprint.py]: Also don't use str(). In Python 2 you'll frequently encounter UnicodeError.

File: oclubs/blueprints/clubblueprint.py [hunk 10]
```diff
@@ -166,40 +178,44 @@ def memberinfo_download(club):
...
+    if desc:
         club.description = FormattedText.handle(current_user, club,
                                                 request.form['description'])
-    if request.files['picture'].filename != '':
+
+    if request.files.getlist('picture'):
         try:
             club.picture = Upload.handle(current_user, club,
-                                         request.files['picture'])
+                                         request.files.getlist('picture')[0])
```
Comment (zhuyifei1999) [inline oclubs/blueprints/clubblueprint.py]: Uh, could you clarify the purpose of this?

File: oclubs/blueprints/resblueprint.py [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+
+    form = ClassroomSidebarForm()
+    # dynamically set the selection form choices
+    form.classrooms_list.choices = classrooms_list
+
+    clearBtn = ClearSelectionForm()
+
+    if request.method == 'POST':
+        # rebuild the res_filter
+        temp = list(res_filter.conds)
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: Rename this variable. 'temp' is very undescriptive.

File: oclubs/blueprints/resblueprint.py [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+
+    rdict = Classroom.get_free_classroom_conditions(
+        **room_filter.to_kwargs())
+
+    # partially unpack dict
+    buildings = [building for building in rdict.keys()]
+    timeslots = [timeslot for timeslot in rdict[buildings[0]].keys()]
+    buildings.sort(key=lambda b: b.format_name)
+    # Order: Lunch, Afterschool
+    timeslots.sort(key=lambda t: t.format_name, reverse=True)
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: This is an Enum right? Can you sort by its numerical value?

File: oclubs/blueprints/resblueprint.py [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+        dates.sort()
+
+        # display order: date, building, timeslot, rooms
+        info = [dates, buildings, timeslots, rdict]
+
+    form = ViewClassroomsForm()
+
+    if request.method == "POST":
+        # rebuild the room_filter
+        temp = list(room_filter.conds)
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: ditto

File: oclubs/blueprints/resblueprint.py [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+    if free_classrooms:
+        choices = [(r.room_number, r.room_number) for r in free_classrooms]
+    else:
+        choices = [('none', 'None available')]
+
+    # update the choices on the client side
+    return jsonify(choices)
+
+
+@resblueprint.route('viewres/club/<club>',  methods=['GET', 'POST'])
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: ditto on prefix /

File: oclubs/blueprints/resblueprint.py [hunk 15]
```diff
@@ -0,0 +1,461 @@
...
+        reserver_club=club)
+
+    canReserveForm = ChangeCanReservationForm(
+        changeCanReserve=str(club.reservation_allowed))
+    canSBForm = ChangeCanSmartboardForm(
+        changeCanSB=str(club.smartboard_allowed))
+
+    if request.method == 'POST':
+        if canReserveForm.submit.data:
+            club.reservation_allowed = (
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: club.reservation_allowed = canReserveForm.changeCanReserve.data == 'True'

File: oclubs/blueprints/resblueprint.py [hunk 16]
```diff
@@ -0,0 +1,461 @@
...
+                True
+                if canReserveForm.changeCanReserve.data == 'True'
+                else False)
+
+        if canSBForm.submit.data:
+            club.smartboard_allowed = (
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: ditto

File: oclubs/blueprints/resblueprint.py [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+@get_callsign_decorator(Reservation, 'reservation')
+def deletereservation(reservation):
+    club = reservation.reserver_club
+    single_date = reservation.date
+    timeslot = reservation.timeslot
+    building = reservation.classroom.building
+    room_number = reservation.classroom.room_number
+    owner = current_user
+
+    if reservation.status == 1:
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: What are these numerical values? Can we use Enum?

File: oclubs/blueprints/resblueprint.py [hunk 18]
```diff
@@ -0,0 +1,461 @@
...
+        reservation.activity.reservation = None
+
+    try:
+        ret = Reservation.delete_reservation(single_date,
+                                             timeslot,
+                                             building,
+                                             room_number,
+                                             owner)
+    except NoRow:
+        abort(500)
```
Comment (zhuyifei1999) [inline oclubs/blueprints/resblueprint.py]: Don't abort(500). that doesn't explain the issue with a Traceback. NoRow will generate a 500 anyways

File: oclubs/blueprints/userblueprint.py [hunk 19]
```diff
@@ -65,13 +65,15 @@ def personal():
                                activities=activities,
                                leader_club=leader_club,
                                allow_club_creation=allow_club_creation,
-                               receive_email=receive_email)
+                               receive_email=receive_email,
+                               is_user=True)
```
Comment (zhuyifei1999) [inline oclubs/blueprints/userblueprint.py:76]: uh, what is this is_user?

File: oclubs/forms/classroom_forms.py [hunk 20]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
+# -*- coding: UTF-8 -*-
+#
+
+from __future__ import absolute_import, unicode_literals, division
+
+from oclubs.utils.dates import today, ONE_DAY, DATE_RANGE_MIN, DATE_RANGE_MAX
+
+from flask_wtf import FlaskForm
+# from wtforms.widgets import ListWidget, CheckboxInput
```
Comment (zhuyifei1999) [inline oclubs/forms/classroom_forms.py]: Is this used? If not please delete the import.

File: oclubs/objs/activity.py [hunk 21]
```diff
@@ -4,45 +4,39 @@
 
 from __future__ import absolute_import, unicode_literals, division
 
-from datetime import datetime, date, timedelta
+from datetime import date, timedelta
 import json
 
+from oclubs.utils.dates import ONE_DAY, int_to_dateobj, dateobj_to_int
```
Comment (zhuyifei1999) [inline oclubs/objs/activity.py:10]: Nice :)

File: oclubs/static-dev/css/content.pcss [hunk 22]
```diff
@@ -0,0 +1,281 @@
+/*
```
Comment (zhuyifei1999) [inline oclubs/static-dev/css/content.pcss:1]: This file is more of a plain CSS than PostCSS:

File: oclubs/static-dev/css/content.pcss [hunk 23]
```diff
@@ -0,0 +1,281 @@
...
+#search_bar_beside_title {
+  margin-top: 20px;
+  margin-bottom: 5px;
+}
+
+.title_link {
+  text-decoration: underline;
+}
+
+.title_link:hover {
```
Comment (zhuyifei1999) [inline oclubs/static-dev/css/content.pcss:26]: &:hover

File: oclubs/static-dev/css/content.pcss [hunk 24]
```diff
@@ -0,0 +1,281 @@
...
+  text-decoration: none !important;
+  cursor: pointer;
+}
+/* Table elements that act as clickable links */
+/* Post things: Jumbotron and well <div>s */
+.club_list_image {
+  width: 100%;
+  height: auto;
+}
+@media (max-width: 2560px) {
```
Comment (zhuyifei1999) [inline oclubs/static-dev/css/content.pcss]: .post_image {
 @media... {}
 @media... {}
}

File: oclubs/static-dev/css/content.pcss [hunk 25]
```diff
@@ -0,0 +1,281 @@
...
+    -webkit-transform: translate(2px, 2px);
+    transform: translate(2px, 2px);
+  }
+}
+
+.lds-magnify {
+  position: relative;
+}
+
+.lds-magnify > div {
```
Comment (zhuyifei1999) [inline oclubs/static-dev/css/content.pcss:247]: & > div

File: oclubs/static-dev/css/content.pcss [hunk 26]
```diff
@@ -0,0 +1,281 @@
...
+  -webkit-transform: scale(0.8);
```
Comment (zhuyifei1999) [inline oclubs/static-dev/css/content.pcss:248]: Don't add prefixes manually. PostCSS handles them

File: oclubs/static-dev/js/jquery.fileupload.js [hunk 27]
```diff
@@ -0,0 +1,1486 @@
+/*
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/jquery.fileupload.js]: Load from CDNJS 'blueimp-file-upload'

File: oclubs/static-dev/js/masonry.js [hunk 28]
```diff
@@ -0,0 +1,2504 @@
+/*!
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/masonry.js]: Load from CDNJS 'masonry'

File: oclubs/static-dev/js/newreservation.js [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+        $('.loading_gif').show();
+        $('.finished_loading').hide();
+      });
+
+      $(document).ajaxStop(function() {
+        $('.loading_gif').hide();
+        $('.finished_loading').show();
+
+
+				Date.prototype.today = function() {
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/newreservation.js]: Inconsistent indentation

Comment (zhuyifei1999) [reply oclubs/static-dev/js/newreservation.js]: Also, having your own methods is preferred then writing to a built-in's prototype

File: oclubs/static-dev/js/oclubs.js [hunk 30]
```diff
@@ -1,159 +1,241 @@
...
+    });
+}(jQuery));
+
+$(document).on('click', '.navbar-collapse.in', function(e) {
+  if ($(e.target).is('a')) {
+    $(this).collapse('hide');
+  }
+});
+
+function show_scroll_to_top_btn() {
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/oclubs.js]: Don't define functions outside IIFEs.

File: oclubs/static-dev/js/select2.js [hunk 31]
```diff
@@ -0,0 +1,5847 @@
+/*!
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/select2.js]: Load from CDNJS 'select2'

File: oclubs/static-dev/js/selection.js [hunk 32]
```diff
@@ -0,0 +1,42 @@
...
+
+// function toggle_date_options_div(){
+// 	if ( $('#viewclassroom_options-0').prop('checked') ){
+// 		$( '#date_options_div' ).hide();
+// 	} else if ( $('#viewclassroom_options-1').prop('checked') ){
+// 		$( '#date_options_div' ).show();
+// 	}
+// }
+
+function toggle_dateselectionform(){
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/selection.js]: Ditto.

File: oclubs/static-dev/js/selection.js [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+
+		$( '#date_options' ).on('change', function(){
+			toggle_dateselectionform();
+		});
+
+	} );
+}( jQuery ) );
+
+
+// function toggle_date_options_div(){
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/selection.js]: Is this still used? If not please delete it.

File: provision/puppet/hieradata/common.yaml [hunk 34]
```diff
@@ -80,7 +79,8 @@ nginx::nginx_vhosts:
     listen_options: default_server
     uwsgi: uwsgi
     add_header:
-      Content-Security-Policy: "default-src \\'self\\'"
+      Content-Security-Policy: "default-src \\'self\\'; img-src \\'self\\' https://loremflickr.com"
```
Comment (zhuyifei1999) [inline provision/puppet/hieradata/common.yaml]: No. Either Proxy or host yourself.

**Review by zhuyifei1999 — CHANGES_REQUESTED** [review]
Good job on Python changes.
The newer PostCSS code are more like pure CSS than PostCSS. 
JS needs eslint. The styles are very inconsistent.
No comment on HTML changes.

File: oclubs/blueprints/userblueprint.py (hunk 19 above)
Comment (AngleQian) [reply oclubs/blueprints/userblueprint.py:76]: same as is_list, is_act, etc. This is to appropriately highlight the nav bar.

File: oclubs/static-dev/css/content.pcss (hunk 25 above)
Comment (AngleQian) [reply oclubs/static-dev/css/content.pcss:247]: This section of css is generated from a loading animation generation site.

Comment (zhuyifei1999) [reply oclubs/static-dev/css/content.pcss:247]: Then convert it

File: oclubs/blueprints/userblueprint.py (hunk 19 above)
Comment (zhuyifei1999) [reply oclubs/blueprints/userblueprint.py:76]: oh ok.

File: oclubs/objs/classroom.py [hunk 35]
```diff
@@ -0,0 +1,174 @@
...
+                                 timeslot=None, order_by_room_number=True,
+                                 pager=None):
+
+        """
+        Get classrooms （regardless of whether it is reserved)
+
+        buildings type: either one Building object or list of Building objects
+        timeslot type: ActivityTime object
+
+        return: list of Classroom objects
```
Comment (zhuyifei1999) [inline oclubs/objs/classroom.py]: @returns:

This is Sphinx.

File: oclubs/static-dev/js/oclubs.js [hunk 36]
```diff
@@ -144,16 +170,97 @@
...
+					document.getElementById( 'scroll_to_top_btn' )
+						.style.display = 'block';
+				} else {
+					document.getElementById( 'scroll_to_top_btn' )
+						.style.display = 'none';
+				}
+			}
+
+			window.onscroll = function () {
+				showScrollToTopBtn();
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/oclubs.js]: Is this the only use of `showScrollToTopBtn`. If so, could you inline the function?

File: oclubs/static-dev/js/oclubs.js [hunk 37]
```diff
@@ -144,16 +170,97 @@
...
+
+			function toggleDayNightBtn( currentClass ) {
+				var display = currentClass === 'day-mode' ? 'Night theme' : 'Day theme';
+				$( '#day-night-toggle' )
+					.html( display );
+			}
+
+			function toggleDayNight() {
+				// set body class
+				var body = document.getElementById( 'body' );
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/oclubs.js]: $( 'body' )

File: oclubs/static-dev/js/oclubs.js [hunk 38]
```diff
@@ -144,16 +170,97 @@
...
+				var currentClass = body.className;
+
+				body.className = currentClass === 'day-mode' ? 'night-mode' : 'day-mode';
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/oclubs.js]: Use .toggleClass(). An element may have multiple classes applied and you shouldn't judge based on the full class expression.

File: oclubs/static-dev/js/oclubs.js [hunk 39]
```diff
@@ -144,16 +170,97 @@
...
+					.className = localStorage.theme;
+			} else {
+				localStorage.theme = 'day-mode';
+			}
+
+			toggleDayNightBtn( localStorage.theme );
+
+			$( '#day-night-toggle' )
+				.click( function () {
+					event.preventDefault();
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/oclubs.js]: is `event` defined?

File: oclubs/static-dev/js/oclubs.js [hunk 40]
```diff
@@ -144,16 +170,97 @@
...
+			if ( localStorage.theme ) {
+				document.getElementById( 'body' )
+					.className = localStorage.theme;
+			} else {
+				localStorage.theme = 'day-mode';
+			}
+
+			toggleDayNightBtn( localStorage.theme );
+
+			$( '#day-night-toggle' )
```
Comment (zhuyifei1999) [inline oclubs/static-dev/js/oclubs.js]: This is a JS-only functionality. Make it not show up when JS is disabled

**Review by zhuyifei1999 — APPROVED** [review]

**Pull request merged**